h2/src/main/org/h2/mvstore/tx/TxDecisionMaker.java

```diff
@@ -270,31 +270,25 @@ public final String toString() {
                 long id = existingValue.getOperationId();
                 if (id == 0 // entry is a committed one
                             // or it came from the same transaction
-                        || isThisTransaction(blockingId = TransactionStore.getTransactionId(id))) {
-                    if(existingValue.getCurrentValue() != null) {
-                        return decideToAbort(existingValue.getCurrentValue());
+                        || isThisTransaction(blockingId = TransactionStore.getTransactionId(id))
+                            // or entry belongs to a committing transaction
+                            // and therefore will be committed soon
+                        || isCommitted(blockingId)) {
+                    // check what transaction have been doing and fail it wasn't an entry removal
+                    V currentValue = existingValue.getCurrentValue();
+                    if(currentValue != null) {
+                        return decideToAbort(currentValue);
                     }
-                    if (id == 0) {
+                    if (id == 0 || !isThisTransaction(TransactionStore.getTransactionId(id))) {
+                        // even if that commit will result in entry removal
+                        // current operation should fail within repeatable read transaction
+                        // if initial snapshot carries some value
                         V snapshotValue = getValueInSnapshot();
                         if (snapshotValue != null) {
                             return decideToAbort(snapshotValue);
                         }
                     }
                     return logAndDecideToPut(existingValue, existingValue.getCommittedValue());
-                } else if (isCommitted(blockingId)) {
-                    // entry belongs to a committing transaction
-                    // and therefore will be committed soon
-                    if(existingValue.getCurrentValue() != null) {
-                        return decideToAbort(existingValue.getCurrentValue());
-                    }
-                    // even if that commit will result in entry removal
-                    // current operation should fail within repeatable read transaction
-                    // if initial snapshot carries some value
-                    V snapshotValue = getValueInSnapshot();
-                    if (snapshotValue != null) {
-                        return decideToAbort(snapshotValue);
-                    }
-                    return logAndDecideToPut(null, null);
```

Comment:
I noticed that `(null, null)` is changed to `(existingValue, existingValue.getCommittedValue())` here, but I have no idea about impact of this change and cannot say is it valid or not.

Reply:
I was aware of that, but I thought it's ok, and values are going to be null either way.
I do not know what I've been smoking at the time :smiley_cat:, but now I realized that I was 
totally wrong.
Unfortunately this logic is not really covered by tests, because you have to do it while another transaction is committing and then roll back your transaction to see the difference.
Will think of other ways to de-dup the code.